---
title: "first"
output: pdf_document
date: "2025-04-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## CHECK DIFFERENTIAL EXPRESSION 
 
```{r Load library}
library(SeuratObject)
library(Seurat)
library(ggplot2)
library(SingleCellExperiment)
library(tidyr)
```
 
Using obj.RData from /pipseq/data which is from line 0 to line 212 from Ellie's notebook include all the data from Single Cell folder

```{r}
#check meta data(data that describes other data) columns
colnames(obj@meta.data)
```
```{r}
#check few rows of meta data result
head(obj@meta.data)
```
```{r}
# View the cluster identity assigned to the first few cells in the Seurat object.
Idents(obj) %>% head()

# Display all the unique cluster levels currently present in the object.
levels(Idents(obj))
```
```{r}
#this renaming part copy from Ellie's code, assume all correct, find cluster of microglia and astrocytes

UMAP_cluster_ids <- c(
  "0" = "Oligodendrocytes",
  "1" = "L2/3IT",
  "2" = "L5IT",
  "3" = "L6IT",
  "4" = "L5ET",
  "5" = "Microglia",
  "6" = "Astrocytes", 
  "7" = "L6CT",
  "8" = "VIP/Lamp5IN",
  "9" = "L2/3IT",
  "10" = "SST/PVIN",
  "11" = "Oligodendrocytes",
  "12" = "OPC",
  "13" = "L5/6NP",
  "14" = "Microglia",
  "15" = "Astrocytes", 
  "16" = "VLMC",
  "17" = "Endo",
  "18" = "VLMC",
  "19" = "Pericytes", 
  "20" = "L2/3IT",
  "21" = "Microglia"
)
Idents(obj) <- obj$seurat_clusters
names(UMAP_cluster_ids) <- levels(obj)
obj <- RenameIdents(obj, UMAP_cluster_ids)
```

```{r}
DimPlot(
  obj,
  reduction = "umap.unintegrated",
  group.by = "condition",
  label = FALSE, # Adds the subclass name to the plot
  label.size = 4 # Adjusts the font size of the labels
)
```

```{r}
#check if renaming works
Idents(obj) %>% head()
levels(Idents(obj))
```
```{r}
table(Idents(obj))
```

```{r}
#group clusters by cell types/cell subtypes
group_cluster <- function(cell_types){
  subset(obj, idents = cell_types)
}
cell_type_list <- c("Astrocytes", "Microglia", "L5IT", "L6IT", "L5ET", "L6CT", "L2/3IT" )

#store all subtypes
cell_subsets <- list()

# for loop for running the list
for (ct in cell_type_list) {
  cell_subsets[[ct]] <- group_cluster(ct)
}

```

```{r}
# Subset subtype to only include those with fewer than 3000 detected genes
filter_gene <- function(subtype) {
  subset(subtype, nFeature_RNA < 3000)
}

for (name in names(cell_subsets)) {
  cell_subsets[[name]] <- filter_gene(cell_subsets[[name]])
}

```


```{r}
cell_subsets[["Microglia"]]

```
```{r}
#normalize data
normalize_data <- function(subtype){
  subtype <- NormalizeData(subtype)
  subtype <- FindVariableFeatures(subtype)
  subtype <- ScaleData(subtype)
  subtype <- RunPCA(subtype)
  subtype <- FindNeighbors(subtype, dims = 1:15)
  subtype <- FindClusters(subtype, resolution = 0.3)
  subtype <- RunUMAP(subtype, dims = 1:15)
  return(subtype)
}

for (name in names(cell_subsets)) {
  cell_subsets[[name]] <- normalize_data(cell_subsets[[name]])
}

```


```{r}
#find marker function
find_marker <- function(seurat_obj, condition1, condition2, group.by = NULL) {
  if (!is.null(group.by)) {
    Idents(seurat_obj) <- seurat_obj[[group.by]][,1]  # set identity from metadata column
  }

  markers <- FindMarkers(
    seurat_obj,
    ident.1 = condition1,
    ident.2 = condition2,
    logfc.threshold = 0.25,
    min.pct = 0.1,
    test.use = "wilcox"
  )

  return(markers)
}
```

```{r}
unique(cell_subsets[["Astrocytes"]]@meta.data$orig.ident)

```


```{r}
library(ggplot2)
library(ggrepel)

plot_volcano <- function(deg, title){
  deg$gene <- rownames(deg)
  deg$log10p <- -log10(deg$p_val_adj)
  deg$log10p[is.infinite(deg$log10p)] <- NA  # handle p_val_adj == 0
  
  # Cap log10p to prevent extreme spikes
  deg$log10p_capped <- pmin(deg$log10p, 50)
  
  # Determine significance
  deg$significance <- ifelse(
    deg$p_val_adj < 0.05 & abs(deg$avg_log2FC) > 0.25,
    "Significant", "Not significant"
  )
  
  # Label top 20 most "important" DEGs based on |logFC| * logP
  deg$label <- NA
  score <- deg$log10p * abs(deg$avg_log2FC)
  top20 <- order(score, decreasing = TRUE)[1:20]
  deg$label[top20] <- deg$gene[top20]
  
  #plot
  ggplot(deg, aes(x = avg_log2FC, y = log10p_capped)) +
    geom_point(aes(color = significance), alpha = 0.5, size = 1.5) +
    geom_text_repel(aes(label = label), size = 3, max.overlaps = 30, box.padding = 0.5) +
    geom_vline(xintercept = c(-0.25, 0.25), linetype = "dotted", color = "black") +
    geom_hline(yintercept = -log10(0.05), linetype = "dotted", color = "black") +
    scale_color_manual(values = c("Significant" = "firebrick", "Not significant" = "grey80")) +
    labs(
      title = title,
      x = "Log2 Fold Change",
      y = "-log10 Adjusted P-value",
      color = "Significance"
    ) +
    theme_minimal(base_size = 13) +
    theme(
      plot.title = element_text(hjust = 0.5),
      legend.position = "right"
  )

}
```


```{r}
library(pheatmap)

plot_heatmap <- function(deg, seurat_obj, condition_column, condition1, condition2, top_n = 20) {
  # Subset Seurat object to the two conditions
  seurat_subset <- subset(seurat_obj, cells = colnames(seurat_obj)[seurat_obj[[condition_column]][,1] %in% c(condition1, condition2)])
  
  # Extract expression matrix (normalized data)
  expr_matrix <- GetAssayData(seurat_subset, layer = "data")
  
  # Get top DEGs
  top_genes <- head(rownames(deg[order(deg$p_val_adj), , drop = FALSE]), top_n)
  
  # Subset expression matrix to top genes
  heatmap_data <- expr_matrix[top_genes, , drop = FALSE]
  heatmap_data <- as.matrix(heatmap_data)
  heatmap_data <- heatmap_data[complete.cases(heatmap_data), ]
  
  # Make column annotation
  annotation_col <- data.frame(
    Condition = seurat_subset@meta.data[[condition_column]]
  )
  rownames(annotation_col) <- colnames(seurat_subset)
  
  # Plot
  pheatmap(heatmap_data,
           cluster_rows = TRUE,
           cluster_cols = FALSE,
           scale = "row",
           show_rownames = TRUE,
           show_colnames = FALSE,
           fontsize_row = 8,
           annotation_col = annotation_col)
}
```

```{r}
add_group_combined <- function(obj) {
  obj$group_combined <- ifelse(obj$orig.ident %in% c("tambexA", "tambexB", "tambexC"), "tambex",
                         ifelse(obj$orig.ident %in% c("vehA", "vehB", "vehC"), "veh",
                         ifelse(obj$orig.ident %in% c("naiveB", "naiveC"), "naive",
                         ifelse(obj$orig.ident %in% c("cpzB", "cpzC"), "cpz", obj$orig.ident))))
  return(obj)
}

for (name in names(cell_subsets)) {
  cell_subsets[[name]] <- add_group_combined(cell_subsets[[name]])
}

```

```{r}
comparison_list <- list(
  c("tambex1wk", "veh1wk"),
  c("tambex", "veh"),
  c("naive", "cpz")
)
```


```{r}

for (cell_type in cell_type_list) {
  for (comp in comparison_list) {
    condition1 <- comp[1]
    condition2 <- comp[2]
    message(sprintf("Processing: %s | %s vs %s", cell_type, condition1, condition2))

    # Differential expression
    deg <- find_marker(
      seurat_obj = cell_subsets[[cell_type]],
      condition1 = condition1,
      condition2 = condition2,
      group.by = "group_combined"
    )

    # Volcano plot
    p <- plot_volcano(deg, title = paste(cell_type, ":", condition1, "vs", condition2))
    print(p)

  }
}
```

