---
title: "first"
output: pdf_document
date: "2025-04-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## CHECK DIFFERENTIAL EXPRESSION ASTROCYTES AND MICROGLIA
 
```{r Load library}
library(SeuratObject)
library(Seurat)
library(ggplot2)
library(SingleCellExperiment)
library(tidyr)
```
 
Using obj.RData from /pipseq/data which is from line 0 to line 212 from Ellie's notebook include all the data from Single Cell folder

```{r}
#check meta data(data that describes other data) columns
colnames(obj@meta.data)
```
```{r}
#check few rows of meta data result
head(obj@meta.data)
```
```{r}
# View the cluster identity assigned to the first few cells in the Seurat object.
Idents(obj) %>% head()

# Display all the unique cluster levels currently present in the object.
levels(Idents(obj))
```
```{r}
#this renaming part copy from Ellie's code, assume all correct, find cluster of microglia and astrocytes

UMAP_cluster_ids <- c(
  "0" = "Oligodendrocytes",
  "1" = "L2/3IT",
  "2" = "L5IT",
  "3" = "L6IT",
  "4" = "L5ET",
  "5" = "Microglia",
  "6" = "Astrocytes", 
  "7" = "L6CT",
  "8" = "VIP/Lamp5IN",
  "9" = "L2/3IT",
  "10" = "SST/PVIN",
  "11" = "Oligodendrocytes",
  "12" = "OPC",
  "13" = "L5/6NP",
  "14" = "Microglia",
  "15" = "Astrocytes", 
  "16" = "VLMC",
  "17" = "Endo",
  "18" = "VLMC",
  "19" = "Pericytes", 
  "20" = "L2/3IT",
  "21" = "Microglia"
)
Idents(obj) <- obj$seurat_clusters
names(UMAP_cluster_ids) <- levels(obj)
obj <- RenameIdents(obj, UMAP_cluster_ids)
```

```{r}
DimPlot(
  obj,
  reduction = "umap.unintegrated",
  group.by = "condition",
  label = FALSE, # Adds the subclass name to the plot
  label.size = 4 # Adjusts the font size of the labels
)
```

```{r}
#check if renaming works
Idents(obj) %>% head()
levels(Idents(obj))
```
```{r}
table(Idents(obj))
```

```{r}
#group clusters 
microglia  <- subset(obj, idents = "Microglia")
astrocytes <- subset(obj, idents = "Astrocytes")
```

```{r}
# Subset microglia and astrocytes to only include those with fewer than 3000 detected genes
microglia <- subset(microglia, nFeature_RNA < 3000)
table(microglia$orig.ident)
astrocytes <- subset(astrocytes, nFeature_RNA < 3000)
table(astrocytes$orig.ident)
```
```{r}
Assays(microglia)
Assays(astrocytes)
```
```{r}
#create seurat object by using slot = data
microglia <- CreateSeuratObject(counts = GetAssayData(microglia, layer = "data"))
astrocytes <- CreateSeuratObject(counts = GetAssayData(astrocytes, layer = "data"))
```

```{r}
#Normalize microglia
microglia <- NormalizeData(microglia)
microglia <- FindVariableFeatures(microglia)
microglia <- ScaleData(microglia)
microglia <- RunPCA(microglia)
microglia <- FindNeighbors(microglia, dims = 1:15)
microglia <- FindClusters(microglia, resolution = 0.3)
microglia <- RunUMAP(microglia, dims = 1:15)
```

```{r}
#Normalize astrocytes
astrocytes <- NormalizeData(astrocytes)
astrocytes <- FindVariableFeatures(astrocytes)
astrocytes <- ScaleData(astrocytes)
astrocytes <- RunPCA(astrocytes)
astrocytes <- FindNeighbors(astrocytes, dims = 1:15)
astrocytes <- FindClusters(astrocytes, resolution = 0.3)
astrocytes <- RunUMAP(astrocytes, dims = 1:15)
```

```{r}
#cell type label for each subset
astrocytes$celltype <- "Astrocytes"
microglia$celltype <- "Microglia"
```

```{r}
#merges the astrocytes and microglia Seurat objects into a single object
combined_obj <- merge(astrocytes, y = microglia)
Idents(combined_obj) <- combined_obj$celltype
```

```{r}
#compares astrocytes vs. microglia to identify differentially expressed genes
astro_micro_de <- FindMarkers(
  combined_obj,
  ident.1 = "Astrocytes",
  ident.2 = "Microglia",
  logfc.threshold = 0.25,    
  min.pct = 0.1,              
  test.use = "wilcox"        
)
```
```{r}
astro_micro_de_sorted <- astro_micro_de[order(astro_micro_de$p_val_adj), ]
head(astro_micro_de_sorted, 20) 
```
```{r}
write.csv(astro_micro_de, "differential_expression_astro_micro/astro_vs_microglia_DE.csv", row.names = TRUE)
write.csv(astro_micro_de, "differential_expression_astro_micro/astro_micro_de_sorted.csv", row.names = TRUE)

```

```{r}
# Read the full differential expression result
one <- read.csv("differential_expression_astro_micro/astro_vs_microglia_DE.csv", row.names = 1)

# Read the top 20 sorted genes
two <- read.csv("differential_expression_astro_micro/astro_micro_de_sorted.csv", row.names = 1)
```

